libcontainer/cgroups/systemd/common.go

```diff
@@ -295,15 +298,26 @@ func generateDeviceProperties(rules []*devices.Rule) ([]systemdDbus.Property, er
 // and returns it
 func getDbusConnection(rootless bool) (*systemdDbus.Conn, error) {
 	connOnce.Do(func() {
-		if rootless {
-			connDbus, connErr = NewUserSystemdDbus()
-		} else {
-			connDbus, connErr = systemdDbus.New()
-		}
+		isRootless = rootless
+		connDbus, connErr = newDbusConnection(rootless)
 	})
 	return connDbus, connErr
 }
 
+func renewDbusConnection() {
+	if atomic.CompareAndSwapInt32(&isRenewing, 0, 1) {
```

Comment:
Too bad there's no higher-level primitive in golang for this.

Makes sense at least add a comment.

Reply:
> If you run this code with go build -race it almost certainly will trigger race warnings.

As of https://github.com/opencontainers/runc/pull/2799 we already build runc and unit tests with -race enabled (and run all tests). I guess it makes sense to add a test case to test the renew functionality (not quite sure how to do that though, will take a look later).